thentos-core/src/Thentos/Transaction.hs

```diff
@@ -268,6 +269,93 @@ unregisterUserFromService uid sid = void $
     execT [sql| DELETE FROM user_services WHERE uid = ? AND sid = ? |] (uid, sid)
 
 
+-- * persona and context
+
+-- | Add a new persona to the DB. A persona has a unique name and a user to which it belongs.
+-- The 'PersonaId' is assigned by the DB. May throw 'NoSuchUser' or 'PersonaNameAlreadyExists'.
+addPersona :: ST -> UserId -> ThentosQuery e Persona
+addPersona name uid = do
+    res <- queryT [sql| INSERT INTO personas (name, uid) VALUES (?, ?) RETURNING id |]
+                  (name, uid)
+    case res of
+        [Only persId] -> return $ Persona persId name uid
+        _             -> impossible "addContext didn't return a single ID"
+
+-- | Delete a persona. Throw 'NoSuchPersona' if the persona does not exist in the DB.
+deletePersona :: PersonaId -> ThentosQuery e ()
+deletePersona persId = do
+    rows <- execT [sql| DELETE FROM personas WHERE id = ? |] (Only persId)
+    case rows of
+        1 -> return ()
+        0 -> throwError NoSuchPersona
+        _ -> impossible "deletePersona: unique constraint on id violated"
+
+-- | Add a new context. The first argument identifies the service to which the context belongs.
+-- May throw 'NoSuchService' or 'ContextNameAlreadyExists'.
+addContext :: ServiceId -> ContextName -> ContextDescription -> ProxyUri -> ThentosQuery e Context
+addContext ownerService name desc url = do
+    res <- queryT [sql| INSERT INTO contexts (owner_service, name, description, url)
+                        VALUES (?, ?, ?, ?)
+                        RETURNING id |]
+                  (ownerService, name, desc, url)
+    case res of
+        [Only cxtId] -> return $ Context cxtId ownerService name desc url
+        _            -> impossible "addContext didn't return a single ID"
+
+-- | Delete a context. Throw an error if the context does not exist in the DB.
+deleteContext :: ContextId -> ThentosQuery e ()
+deleteContext cxtId = do
+    rows <- execT [sql| DELETE FROM contexts WHERE id = ? |] (Only cxtId)
+    case rows of
+        1 -> return ()
+        0 -> throwError NoSuchContext
+        _ -> impossible "deleteContext: unique constraint on id violated"
+
+-- Connect a persona with a context. Throws an error if the persona is already registered for the
+-- context or if the user has any *other* persona registered for the context. (As we currently
+-- allow only one persona per user and context.)
+registerPersonaWithContext :: Persona -> ContextId -> ThentosQuery e ()
+registerPersonaWithContext persona cxtId = do
+    -- Check that user has no registered personas yet
+    res <- queryT [sql| SELECT count(*)
+                        FROM personas pers, personas_per_context pc
+                        WHERE pers.id = pc.persona_id AND pers.uid = ? AND pc.context_id = ? |]
+                  (Only $ persona ^. personaUid)
```

Comment:
I think the `context_id = ?` is too much?  Or how are there two `?` and only one argument?